libs/sdk-backend-base/src/ldmFactories/dashboard/widgetFactory.ts

```diff
@@ -0,0 +1,63 @@
+// (C) 2019-2021 GoodData Corporation
+// import invariant from "ts-invariant";
+import { ObjRef } from "@gooddata/sdk-model";
+import { IWidget, IDashboardFilterReference } from "@gooddata/sdk-backend-spi";
+import { Builder, IBuilder, resolveValueOrUpdateCallback, ValueOrUpdateCallback } from "../builder";
+
+/**
+ * Common widget props builder
+ *
+ * @alpha
+ */
+export interface IWidgetBaseBuilder<T extends IWidget> extends IBuilder<T> {
+    title(valueOrUpdateCallback: ValueOrUpdateCallback<string>): this;
+    description(valueOrUpdateCallback: ValueOrUpdateCallback<string>): this;
+    ignoreDashboardFilters(valueOrUpdateCallback: ValueOrUpdateCallback<IDashboardFilterReference[]>): this;
+    dateDataSet(valueOrUpdateCallback: ValueOrUpdateCallback<ObjRef | undefined>): this;
+    ref(valueOrUpdateCallback: ValueOrUpdateCallback<ObjRef>): this;
+    id(valueOrUpdateCallback: ValueOrUpdateCallback<string>): this;
+    uri(valueOrUpdateCallback: ValueOrUpdateCallback<string>): this;
+}
+
+/**
+ * @alpha
+ */
+export class WidgetBaseBuilder<T extends IWidget> extends Builder<T> implements IWidgetBaseBuilder<T> {
+    protected setWidget = (updateCallback: (widget: Partial<T>) => Partial<T>): this => {
+        this.item = updateCallback(this.item);
+        return this;
+    };
+
+    protected setWidgetProp = <K extends keyof T>(
+        prop: K,
```

Comment:
Is the generics needed here? Can the prop just have type `keyof T`?

Reply:
Yes, without this, you could assign a value with an incorrect type to the prop (key) - this is reason why there is T[K] check.